tls/s2n_crl.c

```diff
@@ -86,3 +86,133 @@ int s2n_crl_get_issuer_hash(struct s2n_crl *crl, uint64_t *hash) {
+    }
+
+    *status = FINISHED;
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_crl_handle_lookup_callback_result(struct s2n_x509_validator *validator) {
+    RESULT_ENSURE_REF(validator);
+
+    crl_lookup_callback_status status = 0;
+    RESULT_GUARD(s2n_crl_get_lookup_callback_status(validator, &status));
+
+    switch (status) {
+        case FINISHED:
+            validator->state = READY_TO_VERIFY;
+            return S2N_RESULT_OK;
+        case AWAITING_RESPONSE:
+            validator->state = AWAITING_CRL_CALLBACK;
+            RESULT_BAIL(S2N_ERR_ASYNC_BLOCKED);
+        default:
+            RESULT_BAIL(S2N_ERR_INVALID_CERT_STATE);
+    }
+}
+
+S2N_RESULT s2n_crl_invoke_lookup_callbacks(struct s2n_connection *conn, struct s2n_x509_validator *validator) {
+    RESULT_ENSURE_REF(validator);
+    RESULT_ENSURE_REF(validator->cert_chain_from_wire);
+
+    int cert_count = sk_X509_num(validator->cert_chain_from_wire);
+    DEFER_CLEANUP(struct s2n_array *crl_lookup_list = s2n_array_new_with_capacity(sizeof(struct s2n_crl_lookup),
+            cert_count), s2n_array_free_p);
+    RESULT_ENSURE_REF(crl_lookup_list);
+
+    for (int i = 0; i < cert_count; ++i) {
+        struct s2n_crl_lookup * lookup = NULL;
+        RESULT_GUARD(s2n_array_pushback(crl_lookup_list, (void**) &lookup));
+
+        X509 *cert = sk_X509_value(validator->cert_chain_from_wire, i);
+        RESULT_ENSURE_REF(cert);
+        lookup->cert = cert;
+        lookup->cert_idx = i;
+    }
+
+    validator->crl_lookup_list = crl_lookup_list;
+    ZERO_TO_DISABLE_DEFER_CLEANUP(crl_lookup_list);
+
+    /* Invoke the crl lookup callbacks after the crl_lookup_list is stored on the validator. This ensures that if a
+     * callback fails, the memory for all other callbacks that may still be running remains allocated */
+    uint32_t num_lookups = 0;
+    RESULT_GUARD(s2n_array_num_elements(validator->crl_lookup_list, &num_lookups));
+    for (uint32_t i = 0; i < num_lookups; i++) {
+        struct s2n_crl_lookup *lookup = NULL;
+        RESULT_GUARD(s2n_array_get(validator->crl_lookup_list, i, (void**) &lookup));
+        RESULT_ENSURE_REF(lookup);
+
+        int result = conn->config->crl_lookup_cb(lookup, conn->config->crl_lookup_ctx);
+        RESULT_ENSURE(result == S2N_SUCCESS, S2N_ERR_CANCELLED);
+    }
+
+    return S2N_RESULT_OK;
+}
+
+int s2n_crl_lookup_get_cert_issuer_hash(struct s2n_crl_lookup *lookup, uint64_t *hash) {
+    POSIX_ENSURE_REF(lookup->cert);
+    POSIX_ENSURE_REF(hash);
+
+    unsigned long temp_hash = X509_issuer_name_hash(lookup->cert);
+    POSIX_ENSURE(temp_hash != 0, S2N_ERR_INTERNAL_LIBCRYPTO_ERROR);
+
+    *hash = temp_hash;
+
+    return S2N_SUCCESS;
+}
+
+int s2n_crl_lookup_set(struct s2n_crl_lookup *lookup, struct s2n_crl *crl) {
+    POSIX_ENSURE_REF(lookup);
+    POSIX_ENSURE_REF(crl);
+    lookup->crl = crl;
+    lookup->status = FINISHED;
+    return S2N_SUCCESS;
+}
+
+int s2n_crl_lookup_ignore(struct s2n_crl_lookup *lookup) {
```

Comment:
Didn't we want the functionality where a user could say "I only want to check this certificate's CRL, and ignore the rest"? Is that behavior not occurring in this PR?

Reply:
Yes, there will be a way to do this by calling an async return function that will entirely skip the revocation check for a certificate. So you could call s2n_crl_lookup_set on the one you want to check, and skip all the rest. But that's in a future PR.